File: examples/go/mega_flow/storage.go
package main

// Interface + struct embedding — Go's idiomatic composition. Taint
// rides through embedded-struct method promotion into the sink.

// Runner — interface implemented by every Repository variant.
type Runner interface {
	Run() int
}

type Repository struct {
	data Envelope
}

func (r *Repository) Cmd() string {
	return r.data.Cmd
}

func (r *Repository) Run() int {
	c := r.Cmd()
	return Execute(c)
}

// AuditedRepository embeds Repository — method promotion exposes Run.
type AuditedRepository struct {
	*Repository
}

func (a *AuditedRepository) Run() int {
	// Embedded-struct promotion preserves taint across the chain.
	return a.Repository.Run()
}

// Persist wraps the tainted payload in a Repository and dispatches
// via the Runner interface.
func Persist(data Envelope) int {
	var repo Runner = &AuditedRepository{Repository: &Repository{data: data}}
	return repo.Run()
}

File: examples/rust/mega_flow/src/storage.rs
use crate::executor;
use crate::Envelope;

// Trait + impl + generics — Rust's idiomatic composition. Taint rides
// through trait dispatch into the sink.

pub trait Runnable {
    fn run(&self) -> String;
}

pub struct Repository {
    data: Envelope,
}

impl Repository {
    pub fn new(data: Envelope) -> Self {
        Self { data }
    }

    fn cmd(&self) -> &str {
        &self.data.cmd
    }
}

impl Runnable for Repository {
    fn run(&self) -> String {
        let c = self.cmd();
        executor::execute(c)
    }
}

// New-type wrapper delegating via trait — preserves taint.
pub struct AuditedRepository(Repository);

impl AuditedRepository {
    pub fn wrap(data: Envelope) -> Self {
        Self(Repository::new(data))
    }
}

impl Runnable for AuditedRepository {
    fn run(&self) -> String {
        // Delegate preserves taint across the trait chain.
        self.0.run()
    }
}

pub fn persist(envelope: Envelope) -> String {
    let repo = AuditedRepository::wrap(envelope);
    repo.run()
}
